tests/src/Interop/StringMarshalling/BSTR/BSTRTestNative.cpp

```diff
@@ -0,0 +1,195 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "platformdefines.h"
+
+WCHAR* strManaged = L("Managed\0String\0");
+size_t   lenstrManaged = 14; // the byte length of strManaged
+
+WCHAR* strReturn = L("a\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0");
+WCHAR* strerrReturn = L("error");
+
+WCHAR* strNative = L(" Native\0String\0");
+size_t lenstrNative = 14; //the byte len of strNative
+
+//Test Method1
+extern "C" BSTR ReturnString()
+{
+    return TP_SysAllocString(strReturn);    
+}
+
+extern "C" BSTR ReturnErrorString()
+{
+    return TP_SysAllocString(strerrReturn);    
+}
+
+//Test Method2
+extern "C" DLL_EXPORT BSTR Marshal_InOut(/*[In,Out]*/BSTR s)
+{
+    //Check the Input
+    size_t len = TP_SysStringByteLen(s);
+
+    if((len != lenstrManaged)||(memcmp(s,strManaged,len)!=0))
+    {
+        printf("Error in Function Marshal_InOut(Native Client)\n");        
+        
+        for(size_t i = 0; i< lenstrManaged;++i)
+            putchar(*(((char *)strManaged)+i));               
+        
+        for(size_t j = 0; j < len; ++j )
+            putchar(*(((char *)s) + j));
+        return ReturnErrorString();
+    }
+
+    //In-Place Change
+    memcpy(s, strNative, len);
+
+    //Return
+    return ReturnString();
+}
+
+
+extern "C" DLL_EXPORT BSTR Marshal_Out(/*[Out]*/BSTR s)
+{
+    s = TP_SysAllocString(strNative);
+    
+    //In-Place Change
+    memcpy(s,strNative, lenstrNative);
+
+    //Return
+    return ReturnString();
+}
+
+
+extern "C" DLL_EXPORT BSTR MarshalPointer_InOut(/*[in,out]*/BSTR *s)
+{    
+    //Check the Input
+    size_t len = TP_SysStringByteLen(*s);
+
+    if((len != lenstrManaged)||(memcmp(*s,strManaged,len)!=0))
+    {
+        printf("Error in Function MarshalPointer_InOut\n");
+        
+        for(size_t i = 0; i< lenstrManaged;++i)
+            putchar(*(((char *)strManaged)+i));
+                
+        for( size_t j = 0; j < len; ++j)
+            putchar(*(((char *)*s) + j));
+        
+        return ReturnErrorString();
+    }
+
+    //Allocate New
+    TP_SysFreeString(*s);
+    *s = TP_SysAllocString(strNative);
+    
+    
+    //Return
+    return ReturnString();
+}
+
+extern "C" DLL_EXPORT BSTR MarshalPointer_Out(/*[out]*/ BSTR *s)
+{
+    *s = TP_SysAllocString(strNative);
+    return ReturnString();
+}
+
+extern "C" DLL_EXPORT int Writeline(char * pFormat, int i, char c, double d, short s, unsigned u)
```

Comment:
Why is this called `Writeline`? There is no writing happening here and it is confusing. Perhaps sum, but then why is it called `pFormat`?